lms/services/lis_result_sourcedid.py

```diff
@@ -0,0 +1,57 @@
+from lms.models import LISResultSourcedId
+
+__all__ = ["LISResultSourcedIdService"]
+
+
+class LISResultSourcedIdService:
+    """Methods for interacting with LISResultSourcedId records."""
+
+    def __init__(self, _context, request):
+        self._db = request.db
+
+    def upsert(self, validated_attrs):
+        """
+        Update an existing record or create a new one if none exists.
+
+        :arg validated_attrs: Valid attributes for associated
+            :class:`lms.models.LISResultSourcedId` record. It is the
+            responsibility of the caller to validate these attributes.
```

Comment:
I'm not sure I'd bother with this comment---it's always the responsibility of the caller to pass valid arguments to services. For example not to pass a number where a string is expected, or not to omit a required argument. We don't document that in every service's docstrings.

The use of a dict here makes it possible to pass in a dict with required keys missing from it and/or with irrelevant keys in it, neither of which would be possible with simple method arguments. So the docstring could perhaps explain what keys the dict must contain and what each key is and its type.

But tbh I'd probably just skip that and  go straight to using a namedtuple instead of a dict---it's impossible to construct a namedtuple with missing or irrelevant fields, and the namedtuple class itself provides a place to document each field's type and meaning.

Reply:
👍 I'm cutting down the comment, and I will do some follow-up work here to see if I can't just wrangle this into a NamedTuple as suggested.